src/dashboard/calc_pnl.py

```diff
@@ -2,13 +2,80 @@
 from __future__ import annotations
 
 import logging
+from decimal import Decimal
 
+import numpy as np
 import pandas as pd
+from eth_typing import ChecksumAddress, HexAddress, HexStr
+from fixedpointmath import FixedPoint
+from web3 import Web3
 
+from src import eth
 from src.dashboard.extract_data_logs import calculate_spot_price
-
-
-def calc_total_returns(pool_config: pd.Series, pool_info: pd.DataFrame, wallet_deltas: pd.DataFrame) -> pd.Series:
+from src.eth.transactions import smart_contract_preview_transaction
+from src.eth_bots.core.environment_config import EnvironmentConfig
+from src.hyperdrive import contract_interface
+
+
+def calc_closeout_pnl(current_wallet: pd.DataFrame, pool_info: pd.DataFrame, env_config: EnvironmentConfig):
+    """Calculate closeout value of agent positions."""
+    web3: Web3 = eth.initialize_web3_with_http_provider(env_config.rpc_url, request_kwargs={"timeout": 60})
+
+    # send a request to the local server to fetch the deployed contract addresses and
+    # all Hyperdrive contract addresses from the server response
+    addresses = contract_interface.fetch_hyperdrive_address_from_url(env_config.artifacts_url)
+    abis = eth.abi.load_all_abis(env_config.abi_folder)
+    contract = contract_interface.get_hyperdrive_contract(web3, abis, addresses)
+
+    # Define a function to handle the calculation for each group
+    def calc_single_closeout(position: pd.DataFrame, min_output: int, as_underlying: bool):
```

Comment:
Lets move this function definition outside of this function. It's complicated enough to be upgraded from a local function definition.